### PR TITLE
Add better support for `ManyToManyField`'s `through` model

### LINKED ISSUE #883: Error typechecking ManyToManyField manager `request.user.groups.all()`
# Bug report

## What's wrong

If I use `request.user.groups.all()` in a view (request being an instance of `django.http.HttpRequest`), this runs fine but mypy complains with

`error: Item "ManyToManyField[Sequence[Group], RelatedManager[Group]]" of "Union[ManyToManyField[Sequence[Group], RelatedManager[Group]], EmptyManager]" has no attribute "all"`

## How is that should be

Typechecking should pass without errors.

## System information

- OS: Ubuntu 20.04
- `python` version: 3.9.10
- `django` version: 4.0.2
- `mypy` version: 0.910
- `django-stubs` version: 1.9.0
- `django-stubs-ext` version: 0.3.1

### LINKED ISSUE #1549: Many-to-Many generated through model are not properly typed
# Bug report

The generated through model for many-to-many is not getting typed properly.

## What's wrong

In the following example:
```python
from typing import reveal_type
from django.db import models


class Topping(models.Model):
    label = models.CharField(max_length=100)


class Pizza(models.Model):
    toppings = models.ManyToManyField(Topping)


for toppings_through in Pizza.toppings.through.objects.all():
    reveal_type(toppings_through)
    reveal_type(toppings_through.pizza)
    reveal_type(toppings_through.topping)
```

 I would expect the revealed types to be:
 * Not sure, but at least `Model`
 * `Pizza`
 * `Topping`

but the result is:
```
demo.py:14: note: Revealed type is "Any"
demo.py:15: note: Revealed type is "Any"
demo.py:16: note: Revealed type is "Any"
```

`reveal_type(Pizza.toppings.through)` gives `Type[django.db.models.base.Model]`, so I’m assuming that’s where we could be more precise.

## System information

- OS:
- `python` version: 3.11.1
- `django` version: 4.2.2
- `mypy` version: 1.3.0
- `django-stubs` version: 4.2.1

### PR DESCRIPTION
**Pull request by flaeppe**
# I have made things!

I've modified how `ManyToManyField` is generic and also made the `ManyToManyDescriptor` generic. This was to get a better description of the `ManyToManyField`, which requires less plugin interference and results in more accurate types.

The most important thing to notice here is a call without any explicit `through` argument makes the `ManyToManyField` instance only partially declared. This is where the plugin steps in and will detect the call argument is missing and generate a through model definition that aligns with what Django describes under the `ManyToManyField.through` section in their documentation. Here: https://docs.djangoproject.com/en/dev/ref/models/fields/#django.db.models.ManyToManyField.through.

`ManyToManyField.through` should now be properly resolved to either an implicitly generated through model, or an explicitly passed one. For example, expanding a generated through model looks like below:

```python
class MyModel_m2mfield(models.Model):
    id: <DefaultAutoField>
    mymodel: ForeignKey(MyModel)
    mymodel_id: MyModel.pk
    othermodel: ForeignKey(OtherModel)
    othermodel_id: OtherModel.pk

    objects: models.Manager["MyModel_m2mfield"]
```

`ManyToManyField()` calls with lazy references (e.g. `"myapp.MyModel"`) should also be supported, `"self"` too.

## Related issues

Closes: #1549 
Closes: #883
Refs: #1638
Refs: #1454 

---

An additional note here is that I'm getting a feeling that the there's a similar adjustment to be done for e.g. `ForeignKey` and `OneToOneField` and their descriptors, which could improve their definition in general, while requiring less plugin code.

### CODE REVIEW

File: django-stubs/db/models/fields/related.pyi
Comment (sobolevn) [inline django-stubs/db/models/fields/related.pyi:210]: Can we make it generic on all the things and simplify our plugin a bit?

File: mypy_django_plugin/lib/helpers.py
Comment (sobolevn) [inline mypy_django_plugin/lib/helpers.py]:
```suggestion
def is_abstract_model(model: TypeInfo) -> bool:
```

No reason to make it pos-only

File: mypy_django_plugin/transformers/fields.py
Comment (sobolevn) [inline mypy_django_plugin/transformers/fields.py]:
```suggestion
```

File: mypy_django_plugin/transformers/models.py
Comment (sobolevn) [inline mypy_django_plugin/transformers/models.py:809]: This part of the change seems like the most complex one to me.

File: tests/typecheck/fields/test_related.yml
Comment (sobolevn) [inline tests/typecheck/fields/test_related.yml:282]: Let's have both assertions: with `Member().app` and with `.get()`

File: tests/typecheck/fields/test_related.yml
Comment (sobolevn) [inline tests/typecheck/fields/test_related.yml]: Why is it changed?

**Review by sobolevn — COMMENTED** [review]
This is not a full review, just some questions.

Thanks for your work!

File: tests/typecheck/fields/test_related.yml
Comment (flaeppe) [reply tests/typecheck/fields/test_related.yml]: I was fiddling around with some stuff in the beginning. This should be reverted.

File: django-stubs/db/models/fields/related.pyi
Comment (flaeppe) [reply django-stubs/db/models/fields/related.pyi:210]: Which parts are you referring to?

The plugin only acts on what is `_T` and `_M` here, essentially the `to=` and `through=` argument on `ManyToManyField`.

It does however have to sort some stuff out when there was no explicit `through=` argument passed, as Django automatically generates a through model behind the scenes in that case.

Additionally it supports resolving lazy references e.g. `"myapp.MyModel"` and replaces either `_T` or `_M` with any found model type.

File: mypy_django_plugin/transformers/fields.py
Comment (flaeppe) [inline mypy_django_plugin/transformers/fields.py]:
```suggestion
            else:
```

__Edit:__ Nope, nvm, that's the opposite

File: mypy_django_plugin/transformers/models.py
Comment (flaeppe) [reply mypy_django_plugin/transformers/models.py:809]: Yeah, it's messy. All it does is sort out how `ManyToManyField()` was called;

- Did it have `to=` and/or `through=`?
- Resolve any lazy reference for `to=` or `through=`
- Include the model (typeinfo) that `to=` and/or `through=` pointed to

I tried to encapsulate these results in the named tuple `M2MArguments`, and also to get some form of declarative description of how we can interpret a call to `ManyToManyField`.

e.g. required `to` but optional `through`, while `through` can also be an implicit model:

```python
M2MArguments(..., through=M2MThrough(call_arg=None, model=...))
```

File: django-stubs/db/models/fields/related.pyi
Comment (flaeppe) [reply django-stubs/db/models/fields/related.pyi:210]: > Which parts are you referring to?

What I meant to say here was: which other parts should we make generic to simplify the plugin?

Comment (sobolevn) [reply django-stubs/db/models/fields/related.pyi:210]: `_GT, _ST`? Will it help?

Comment (flaeppe) [reply django-stubs/db/models/fields/related.pyi:210]: No, this changes so that those arguments aren't used at all any more, for `ManyToManyField`. It's just the parent class that wants them.

The generics of the descriptor class does almost all the work now instead (except when `through` is implicit, then the plugin does that part)

Comment (flaeppe) [reply django-stubs/db/models/fields/related.pyi:210]: It's a bit hard to follow, but instead of using `_GT` and `_ST` and let our plugin do _all_ the work, I've updated the descriptor class `ManyToManyDescriptor` to become generic over a model, and that will be the `through=` model for a `ManyToManyField()`.

And if you now look a bit further, you find: `ManyToManyDescriptor.through -> type[_M]` which means that the `through=` argument align with the `MyModel.m2m_field.through` without any plugin work (again, except for special cases of lazy referencing or not passing the `through=` argument, which is all the new plugin code you see in this PR)

Comment (flaeppe) [reply django-stubs/db/models/fields/related.pyi:210]: Corresponding part is the `to=` argument, plugin code will _only_ be running for lazy referencing. Previously this was completely managed by the plugin, now it's instead mypy doing the work.

File: mypy_django_plugin/lib/fullnames.py
Comment (flaeppe) [inline mypy_django_plugin/lib/fullnames.py:41]: For extra clarification: the removal of this line detaches `ManyToManyField` to be handled by the `_GT` and `_ST` generic arguments and the attributes `_pyi_private_set_type` and `_pyi_private_get_type`

File: mypy_django_plugin/transformers/fields.py
Comment (flaeppe) [inline mypy_django_plugin/transformers/fields.py]: Might be worth moving this to some `transformers/manytomany.py`, if possible

File: mypy_django_plugin/transformers/fields.py
Comment (flaeppe) [inline mypy_django_plugin/transformers/fields.py]: I suppose this could be broken out to a type guard helper, if we want?

```python
ModelTypeInfo = NewType("ModelTypeInfo", TypeInfo)
def is_model_info(info: TypeInfo) -> TypeGuard[ModelTypeInfo]:
    return (
        info.metaclass_type is not None
        and info.metaclass_type.type.fullname == fullnames.MODEL_METACLASS_FULLNAME
    )
```

File: mypy_django_plugin/transformers/fields.py
Comment (flaeppe) [inline mypy_django_plugin/transformers/fields.py]: To expand a bit more here; I have an idea that we could also attempt to resolve a `MemberExpr` pointing towards `settings.AUTH_USER_MODEL`. As such it's possible to pick up its lazy reference, something like

```python
if isinstance(expr, MemberExpr) and ...:
    label = django_context.settings.AUTH_USER_MODEL
    # Below here is same as when `isinstance(expr, StrExpr)`
    fullname = django_context.model_class_fullnames_by_label.get(label)
        if fullname is not None:
            model_info = helpers.lookup_fully_qualified_typeinfo(api, fullname)
            if model_info is not None:
                return Instance(model_info, [])
            elif isinstance(api, SemanticAnalyzer) and not api.final_iteration:
                # Getting this far, where Django matched the reference but we still can't
                # find it, we want to defer
                api.defer()
```

Like so, there could be another helper:

```python
def resolve_lazy_reference(expr: Expression) -> Optional[TypeInfo]:
    ...
```

Which focuses on converting any `StrExpr("<app_label>"."<object_name>")` to a `TypeInfo`. This could probably be reused in multiple other places e.g. `Apps.get_model(...)` and `to=` argument for `ForeignKey`, `OneToOneField`, etc.

File: mypy_django_plugin/transformers/fields.py
Comment (flaeppe) [inline mypy_django_plugin/transformers/fields.py]: Hm, I think we could yield an error on the `else` clause here, saying that the lazy reference is invalid and doesn't point to a model

**Comment by flaeppe** [comment]
@sobolevn I know that this is a large addition. I wanted to have it more in pieces but wasn't really able to, I pulled in one end and some stuff started to follow along.

Anyways, I've run this on one of our larger projects, hit a few issues that I resolved and it's now green except 1 part, namely:

```python
users = ManyToManyField(settings.AUTH_USER_MODEL)
```

But I wanted to keep that as a separate change to avoid piling up even more in here.

I'm asking for an opinion here, how could we proceed to make it mergable?

Additionally, @ljodal, I know you've previously had possibility to run changes on projects generating a bit more use cases, could you perhaps try this one out? See if we can touch a bit wider surface? Perhaps there's some cheap fixes to be found before anything is merged.

All in all this PR should unlock some `ManyToManyField`-box, as we haven't really had any good support for it previously. No idea what'll come out of it though. Just hoping that it's not gonna generate too many crashes..

**Comment by ljodal** [comment]
I'll give it a go, we haven't updated django-stubs in a while though, so not sure how much unrelated will pop up

**Comment by flaeppe** [comment]
Additionally I want to emphasize that what I've done here is actually to use __more__ of `mypy` and less plugin.

Starting to use the descriptor class is the trick, it does __half__ of the work when having an implicit `through=` table:

```python
others = ManyToManyField(Other)
```

And it does __all__ the work when having an explicit `through=` table:

```python
others = ManyToManyField(Other, through=InBetween)
```

- Then there's a bunch of plugin code helping out with resolving lazy references: `"app_label.Model"`. But all it does is replace those expressions with model instances, so that `mypy` can see it the same way that I've written it above (☝️).

- The exception is the implicit through table case, where the plugin builds and inserts that class, similar to what Django does. That code looks a bit intimidating but is nothing to be afraid of. It's flat out writing a model class declaration in "mypy code"..

File: mypy_django_plugin/lib/helpers.py
Comment (ljodal) [inline mypy_django_plugin/lib/helpers.py:436]: We have a case of this at least (the referenced model is defined right below in the file), which now causes "Needs type annotation" errors

File: mypy_django_plugin/transformers/manytomany.py
Comment (ljodal) [inline mypy_django_plugin/transformers/manytomany.py:132]: We have a bunch of these cases as well

**Review by ljodal — COMMENTED** [review]
Haven't looked too closely at the code, but it seems to run fine on our largest project. A couple of new warnings due to the limitations mentioned in separate comments, but nothing major

File: mypy_django_plugin/transformers/manytomany.py
Comment (flaeppe) [reply mypy_django_plugin/transformers/manytomany.py:132]: Something tells me this one will hit quite many projects.

**Comment by flaeppe** [comment]
> Haven't looked too closely at the code, but it seems to run fine on our largest project. A couple of new warnings due to the limitations mentioned in separate comments, but nothing major

Thank you for the help! Great to get more input

File: mypy_django_plugin/transformers/manytomany.py
Comment (ljodal) [reply mypy_django_plugin/transformers/manytomany.py:132]: Guess it's fine to do in a separate PR, but maybe make sure to do it before the next release?

Comment (flaeppe) [reply mypy_django_plugin/transformers/manytomany.py:132]: Yeah, that's what I'm thinking too

**Review by sobolevn — APPROVED** [review]
Thanks! But, merging this PR would block the release. Can you please open a new issue with new release planning, so we won't forget about it?

**Comment by flaeppe** [comment]
> Thanks! But, merging this PR would block the release. Can you please open a new issue with new release planning, so we won't forget about it?

I've opened a next release issue: #1740

**Comment by intgr** [comment]
Is this a known issue? When passing argument `through` to `ManyToManyField`: `through=CustomThrough` works, `through="myapp.CustomThrough"` works.

But `through="CustomThrough"` breaks. It should work since the model is defined in the same file. Test patch:

```diff
diff --git a/tests/typecheck/fields/test_related.yml b/tests/typecheck/fields/test_related.yml
index 8f21557e..3ac3117c 100644
--- a/tests/typecheck/fields/test_related.yml
+++ b/tests/typecheck/fields/test_related.yml
@@ -1133,7 +1133,7 @@
                     auto_through = models.ManyToManyField(Other, related_name="autos")
                     # Have multiple M2Ms with implicit through
                     other_again = models.ManyToManyField(Other, related_name="others_again")
-                    custom_through = models.ManyToManyField(Other, through=CustomThrough, related_name="customs")
+                    custom_through = models.ManyToManyField(Other, through="CustomThrough", related_name="customs")
 
 -   case: test_many_to_many_with_lazy_references
     main: |
```

Output:

```
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Invalid output: 
E   Actual:
E     ...
E     main:6: note: Revealed type is "Any"          (diff)
E    main:7: note: Revealed type is "myapp.models.Other"
E     main:8: note: Revealed type is "Any"          (diff)
```

**Comment by ljodal** [comment]
Yes, see https://github.com/typeddjango/django-stubs/pull/1719#discussion_r1336317602 @intgr :)

**Comment by flaeppe** [comment]
Yes, opting out of the lazy reference without "app_label" was intentional, I also implemented the lazy reference resolver for another PR, but had it erroring, and there was a small discussion regarding that:

https://github.com/typeddjango/django-stubs/pull/1725#discussion_r1332599765

Basically it's non-trivial to figure out the correct implicit app label. So I wanted to push it for later. There might perhaps be shortcuts to take when using the runtime model, but I also wanted to minimise touching the runtime.

It's easy to resolve it at least, as it's just to explicitly add the correct app label.

**Pull request merged**
